include/freerdp/utils/chunkstream.h

```diff
@@ -0,0 +1,215 @@
+	 */
+	FREERDP_API ChunkStreamSlot* chunkstream_getStaticMemSlot(ChunkStream* chunkstream,
+	                                                          const BYTE* data, size_t sz);
+
+	/**
+	 * Retrieves a slot containing a memory blob alloced by the user, the chunkstream takes
+	 * ownership of the provided memory, so it will be freed when the chunkstream is destroyed.
+	 * @param chunkstream the chunkstream
+	 * @param data the memory blob to add
+	 * @param sz size of the memory blob
+	 * @return the corresponding ChunkStreamSlot, NULL if it failed (no available slots)
+	 */
+	FREERDP_API ChunkStreamSlot* chunkstream_getMallocSlot(ChunkStream* chunkstream, BYTE* data,
+	                                                       size_t sz);
+
+	/**
+	 * Retrieves a slot from the chunkstream's pool. This slot is taken from the pool with an empty
+	 * used size, so it's usually a good idea to use chunkstreamslot_update_used() to set the space
+	 * used in the chunk.
+	 * @param chunkstream the chunkstream
+	 * @param sz size to retrieve from the pool
+	 * @return the corresponding ChunkStreamSlot, NULL if fails (no available slots, or memory in
+	 * 		 the pool)
+	 */
+	FREERDP_API ChunkStreamSlot* chunkstream_getPoolSlot(ChunkStream* chunkstream, size_t sz);
+
+	/**
+	 * Retrieves a slot from the chunkstream pool and initialize a static stream with it. The
+	 * stream will point on the memory associated with the pool item. It's usually a good idea
+	 * to call chunkstreamslot_update_fromStream() when stream processing is finished, to update
+	 * the used counter of the chunk.
+	 * @param chunkstream the chunkstream
+	 * @param sz size to retrieve from the pool
+	 * @param s the stream to initialize
+	 * @return the corresponding ChunkStreamSlot, NULL if fails (no available slots, or memory in
+	 * 			the pool)
+	 */
+	FREERDP_API ChunkStreamSlot* chunkstream_getPoolStream(ChunkStream* chunkstream, size_t sz,
+	                                                       wStream* s);
+
+	/**
+	 * Assembles all the chunks and put them in the returned wStream
+	 * @param chunkstream the chunkstream
+	 * @return a new wStream containing the assembled chunks
+	 */
+	FREERDP_API wStream* chunkstream_linearizeToStream(ChunkStream* chunkstream);
+
+	/**
+	 * Assemble all the chunks in a provided wStream. The function ensures that there's enough
+	 * remaining space in the target stream.
+	 * @param chunkstream the chunkstream
+	 * @param s the target stream
+	 * @return if the operation was successful
+	 */
+	FREERDP_API BOOL chunkstream_linearizeInStream(ChunkStream* chunkstream, wStream* s);
+
+	/**
+	 * Retrieves the size stored in the chunkstream after a given slot
+	 * @param chunkstream the chunkstream
+	 * @param slot the slot to start to compute
+	 * @return the stored size, -1 if the slot was not found
+	 */
+	FREERDP_API int chunkstream_sizeAfterSlot(const ChunkStream* chunkstream,
```

Comment:
wrong data type, if you require `-1` use `SSIZE_T`

Reply:
ok will change it. Anyway `int` is not problematic, is it ?

Reply:
no, not problematic, but better to use the dedicated size types.